### PR TITLE
[FEATURE] adds undo and redo on transaction groups

### PR DESCRIPTION
**Pull request by vmora**
This work is related to https://github.com/qgis/QGIS-Enhancement-Proposals/issues/97

Only the AddFeature function has been coded for now in order to get some feedback on the chosen direction.

### CODE REVIEW

**Comment by haubourg** [comment]
@m-kuhn does that sound a good start to have the undo redo in QGIS transactions?

**Comment by m-kuhn** [comment]
Yes, this looks very promising, the direction with savepoints sounds very good to me.

Some considerations:

 * Exceptions are not very much used throughout QGIS. Some methods that throw exceptions are overridden Qt methods. Have you verified that all the codepaths that can call these methods safely handle exceptions (without leaking etc)?
 * It might be worthwile to add new virtual methods `QgsTransaction::createSavepoint()`, `QgsTransaction::currentSavepoint()` (and possibly `bool QgsTransaction::currentSavepointValid()` that returns false if an operation took place since the last savepoint)
 * Out of curiosity, how heavy are savepoints (memory- and cpu-wise)?
 * I think we can directly rollback a couple of undo commands in a single step(?). I don't know the performance impact of a single rollback, might be worth detecting such a "fast rollback" and skipping intermediate savepoints, not sure.
 * Use `QgsDebugMsg` instead of `std::cout` (consistency with the rest of the code)
 * Setup your pre-commit hook

**Comment by haubourg** [comment]
For the only part I can answer:

> Out of curiosity, how heavy are savepoints (memory- and cpu-wise)?
It seems to be like a transaction size.
https://serverfault.com/questions/214834/how-much-memory-does-one-postgresql-savepoint-take-up

 When saving or exiting edit mode we need to be sure to emit a `Release SavePoint` . Googling a bit, I found issues on memory consumption, all are very old, and edge cases (JDBC using Latin1 encoding) 
see https://www.spinics.net/lists/pgsql/msg93522.html

@vmora, I let you answering the other good questions :-)

**Comment by vmora** [comment]
@m-kuhn Thanks for the review. Answers follow, but, to make it short, the crucial issue for me is to find a way to handle errors in undo/redo commands. If anyone as an idea, I'll be more than happy to implement/test it.


>Exceptions are not very much used throughout QGIS and Qt. Some methods that throw exceptions are overridden Qt methods.

I'm fully conscious of the fact, and not especially happy with the design. The problem is: the undo/redo is done by the Qt undo-stack, and the redo function has no error mechanism that I could find, while the provider can return an error... hence I resorted to exception mechanism.

> Have you verified that all the codepaths that can call these methods safely handle exceptions (without leaking etc)?

Nope, and I deserve a good ol' flogging for that. The undo (which raises) is connected to a signal, I suppose an uncaught exception there will be trouble, but I'm not sure (http://www.qtcentre.org/archive/index.php/t-44741.html). Under normal circumstances, only the first redo (called by the push) should throw... but if the connection is lost or whatever...

If someone has a clue as to how to handle errors cleanly in a redo, please !

Is the absence of featureAdded signal sufficient ?

> It might be worthwile to add new virtual methods QgsTransaction::createSavepoint() that creates a new savepoint (or returns the current one if nothing happened since the last savepoint?), QgsTransaction::currentSavepoint() (and possibly bool QgsTransaction::currentSavepointValid() that returns false if an operation took place since the last savepoint)

Good idea.

>    I think we can directly rollback a couple of undo commands in a single step(?). I don't know the performance impact of a single rollback, might be worth detecting such a "fast rollback" and skipping intermediate savepoints, not sure.

I don't understand how the "fast roolback" happens on the user side. If you are talking about the insertion of several features (e.g. cpy/paste), then it's a mater of not implementing addFeature**s** as several calls to addFeature. Or are you talking about something else entirely ?

> Use QgsDebugMsg instead of std::cout (consistency with the rest of the code)

Will do my best, but old habits die hard.

> Setup your pre-commit hook

Done. Thanks for the reminder.

**Comment by m-kuhn** [comment]
> If someone has a clue as to how to handle errors cleanly in a redo, please !

In the constructor you receive a pointer to the edit buffer, could you use that to send an error-signal (to at least show it to the user)?
Not sure how to remove the redo command from the stack in this case, but maybe you could somehow "flag" it as failed and leave it on the stack.

> Is the absence of featureAdded signal sufficient ?

I'd say that shouldn't be sent in this case at least, yes. Anything more than that would be nice ot have. @wonder-sk what do you think?

> I don't understand how the "fast roolback" happens on the user side. 

I thought I've seen somewhere a "history" list where a user could directly click 7 steps back, but not sure and this also is just "nice to have" in case 7 savepoint rollbacks are (considerably) slower than a single one.

**Comment by vmora** [comment]
@m-kuhn apparently QUndoCommand::setObsolete should do the trick for command removal, but it's only available since Qt 5.9.

I did start to implement QgsTransaction::createSavepoint()... but revreted. I think it too much hassle for a virtual class that has only one derived class (postgres). If `SAVEPOINT` or `ROOLBACK TO SAVEPOINT` which are standard SQL, are not supported by the next derived class then we may do something more complicated, until then I prefer to keep things minimal.

I removed the use of exceptions, and all should be fine and safe now.

I also changed the behavior of the editPaste function, to use addFeatures instead of addFeature, except for a bigger memory footprint, it should be an improvement overall.

**Comment by vmora** [comment]
@haubourg could you give it a try in a real-world case ?

**Comment by m-kuhn** [comment]
Sounds great, @vmora 

What's the hassle with `QgsTransaction::createSavepoint()`? I was really looking forward to that as I had some ideas what to do with the knowledge of the "current" savepoint.

**Comment by vmora** [comment]
> What's the hassle with `QgsTransaction::createSavepoint()`

- Should it be pure virtual or do nothing by default ? Or shall I implement directly in the base class ?
- Is there the need for a `bool QgsTransaction::supportsSavepoint()` ? I hope not (I like small interfaces a better).
- We need a `QgsTransaction::roolbackToSavepoint(savepoint)`, does it remove the savepoint, or do we also want a `QgsTransaction::removeSavepoint(savepoint)` ? I believe the former is better, in order to write `tr->rollbackToSavepoint( tr->currentSavepoint() ) 
- Having a `currentSavepoint` sorts of duplicate the state/functionality of the undostack, is it what you want ?

The coding is pretty straightforward though, so I can give it a another try and see what it looks like.

**Comment by m-kuhn** [comment]
> Should it be pure virtual or do nothing by default ? Or shall I implement directly in the base class ?

you say it's standard sql? do it in the base class and let the next dev who integrates another provider decide.

> Is there the need for a bool QgsTransaction::supportsSavepoint() ? I hope not (I like small interfaces a better).

dito. let's do this only when required. if ever.

> We need a QgsTransaction::roolbackToSavepoint(savepoint), does it remove the savepoint, or do we also want a QgsTransaction::removeSavepoint(savepoint) ? I believe the former is better, in order to write `tr->rollbackToSavepoint( tr->currentSavepoint() )

not sure. probably leave it and also intermediate ones for redo? unreachable intermediate ones could be removed once a new SP is created on top of an existing one?

> Having a currentSavepoint sorts of duplicate the state/functionality of the undostack, is it what you want ?

I would like to have a "handle" of the current state for other parallel (r/o) connections.

> The coding is pretty straightforward though, so I can give it a another try and see what it looks like.

👍

**Comment by vmora** [comment]
@m-kuhn does this commit look like what you had in mind ?

**Comment by haubourg** [comment]
@m-kuhn Are the more recent commits looking good to you? Can we proceed and extend to update and delete operations?

**Comment by m-kuhn** [comment]
@vmora what I have in mind requires the "current" state as a savepoint. So I imagine a savepoint to be created at the end of each edit command.

If I'm not mistaken, this pull request currently creates a savepoint before starting a new edit command (to backup previous work), is this correct?

**Comment by vmora** [comment]
@m-kuhn you are right, the savepoint is before the edit command, otherwise it'd be a bit complicated to undo the first edit command.

The notion of a 'current' sate makes me think about an undo stack on the transaction side, is that what you have in mind ?

**Comment by m-kuhn** [comment]
> The notion of a 'current' sate makes me think about an undo stack on the transaction side, is that what you have in mind ?

Not sure I understand this. My goal is just to create a second connection and "checkout" the "current" savepoint.

**Comment by vmora** [comment]
> My goal is just to create a second connection and "checkout" the "current" savepoint

Can you really do that ? I can't find a reference for several connection to the same transaction.

**Comment by m-kuhn** [comment]
I don't know yet, but if it works it would be great and at least in theory it sounds feasible :)

**Comment by vmora** [comment]
@m-kuhn ok, I think I got it, I'll try and add save points at the end of the edit command, instead of the beginning.

I can always rollback to the beginning of the transaction to undo the first edit command. So their is always a current save point.

**Comment by vmora** [comment]
@m-kuhn I improved (or I believe I did) the QgsTransaction interface:
```c++
    /**
     * creates a save point
     * returns empty string on error
     * returns the last created savepoint if it's not dirty
     */
    QString createSavepoint( QString &error SIP_OUT );

    /**
     * rollback to save point, the save point is maintained and is "undertied"
     */
    bool rollbackToSavepoint( const QString &name, QString &error SIP_OUT );

    /**
     * dirty save point such that next call to createSavepoint will create a new one
     */
    void dirtyLastSavePoint();
``` 

It'd be nice to call dirtyLastSavePoint() on connection commit(), but it doesn't keep et ref. to the transaction (just a bool to know if there is a transaction), so I dirty the last save point after each commit in the postgres provider.

I also removed the possibility to name a save point in the transaction interface  'cause it uselessly complexified  the class. If you need a named save point you can always executeSQL.

I have **not added** a save point **after** the edit command, if you want me to, I can do that, or you can do that when/if you implement what you had in mind. With the new save point mechanism, it will only create the save point for the next edit command.

I'll implement the remaining undo/redo functions and remove the WIP tag asap.

**Comment by vmora** [comment]
@haubourg, @m-kuhn I have witnessed strange behavior for add/remove fields, maybe they should be "un-undoable" actions ;), what do you think ?

For bulk updates of fields, there is an undo command for each feature, I can see potential problems with that (lots of save points for large datasets). 

Anyway, in the current state, you can test the branch and tell me what you think of it.

**Comment by m-kuhn** [comment]
> I have witnessed strange behavior for add/remove fields, maybe they should be "un-undoable" actions ;), what do you think ?

It'd be better to make them behave nicely 😉

> 
For bulk updates of fields, there is an undo command for each feature, I can see potential problems with that (lots of save points for large datasets).

They should be grouped with editCommandStarted and editCommandEnded.

I hope I find the time to give it a go, but not dure I find it actually.

**Comment by haubourg** [comment]
I build your branch and tested only now; 

So, it works pretty well with a production project (not QGEP) for update / insert and delete. 
I noticed the following : 

- if a transaction failed for any reason, a "action failed" is added to the undo/redo panel. I think we can filter those, since undoing them or redoing them makes no sense

- Redoing a DELETE does not work. I don't see why it couldn't be replayed

- Trying to play with data structure (which I do rarely from QGIS) triggers a direct crash when trying to add a field to a table. Since I have crashes, I can't reproduce your strange behavior !
An unrelated thing, adding a field to a view does not work, QGIS throws an ALTER TABLE. statement. @m-kuhn I remember that in QGEP it worked (which was surprising and led to the action of adding new roles )

I also saw some weird traces in the stdout:

```
ERROR 4: /initrd.img: Permission non accordée
QInotifyFileSystemWatcherEngine::addPaths: inotify_add_watch failed: Aucun fichier ou dossier de ce type
```

**Comment by vmora** [comment]
@haubourg 

> if a transaction failed for any reason, a "action failed" is added to the undo/redo panel. I think we can filter those, since undoing them or redoing them makes no sense

Well, it will be possible with Qt 5.9 ( see above: "apparently QUndoCommand::setObsolete should do the trick for command removal, but it's only available since Qt 5.9.")

> Redoing a DELETE does not work. I don't see why it couldn't be replayed

Works for me (TM)... seriously, I'll look into that.

@m-kuhn 

> They should be grouped with editCommandStarted and editCommandEnded

I'll look into that.

**Comment by m-kuhn** [comment]
> Well, it will be possible with Qt 5.9 ( see above: "apparently QUndoCommand::setObsolete should do the trick for command removal, but it's only available since Qt 5.9.")

Can this be `#ifdef`'d, then it will be available for everyone in a year or two.

File: src/core/qgstransaction.cpp [hunk 1]
```diff
@@ -213,6 +213,19 @@ QString QgsTransaction::createSavepoint( QString &error SIP_OUT )
   return name;
 }
 
+QString QgsTransaction::createSavepoint( const QString &savePointId, QString &error SIP_OUT )
+{
+  if ( !mTransactionActive )
+    return "";
```
Comment (m-kuhn) [inline src/core/qgstransaction.cpp]: `return QString()`

File: src/core/qgstransaction.h [hunk 2]
```diff
@@ -126,6 +126,12 @@ class CORE_EXPORT QgsTransaction : public QObject SIP_ABSTRACT
     QString createSavepoint( QString &error SIP_OUT );
 
     /**
+     * creates a save point
+     * returns empty string on error
```
Comment (m-kuhn) [inline src/core/qgstransaction.h]: `\note Added in QGIS 3.0` but travis will also tell you ;)

**Comment by vmora** [comment]
@haubourg fixed problems with attr add/delete, normally everything works. Could you give it a try plz.

@m-kuhn thanks for the review, I'll have a look latter this afternoon.

**Comment by haubourg** [comment]
@vmora tested this morning, it looks good to me, playing with structure changes and data changes seems stable. 
I think you can proceed with PR's checklist (and add it in the PR description above) : 

- [ ]  Commit messages are descriptive and explain the rationale for changes
- [ ]  Commits which fix bugs include fixes #11111 in the commit message next to the description
- [ ]  Commits which add new features are tagged with [FEATURE] in the commit message
- [ ]  Commits which change the UI or existing user workflows are tagged with [needs-docs] in the commit message and containt sufficient information in the commit message to be documented
- [ ]  I have read the QGIS Coding Standards and this PR complies with them
- [ ]  This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ]  New unit tests have been added for core changes
- [ ]  I have run the scripts/prepare-commit.sh script before each commit

**Comment by vmora** [comment]
@haubourg can you plz try it out once all green and give the go for merge ?

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -7997,13 +7997,18 @@ void QgisApp::editPaste( QgsMapLayer *destinationLayer )
...
-    pasteVectorLayer->addFeature( newFeature );
-    newIds << newFeature.id();
-
+    newFeatures << QgsVectorLayerUtils::createFeature( pasteVectorLayer, geom, dstAttr, &context );
     ++featureIt;
   }
 
+  pasteVectorLayer->addFeatures( newFeatures );
+  QgsFeatureIds newIds;
+  Q_FOREACH ( const QgsFeature &f, newFeatures )
```
Comment (pblottiere) [inline src/app/qgisapp.cpp]: Can you remove the `Q_FOREACH` and use a `for` instead? See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/103

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  }
+  return !hasError();
+}
+
+
+QgsVectorLayerUndoPassthroughCommandAddFeatures::QgsVectorLayerUndoPassthroughCommandAddFeatures( QgsVectorLayerEditBuffer *buffer, QgsFeatureList &features )
+  : QgsVectorLayerUndoPassthroughCommand( buffer, QObject::tr( "add features" ) )
+{
+  static int sAddedIdLowWaterMark = -1;
+  Q_FOREACH ( const QgsFeature &f, features )
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: `Q_FOREACH` -> `for` ?

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    mInitialFeatures.last().setId( sAddedIdLowWaterMark );
+  }
+  mFeatures = mInitialFeatures;
+}
+
+void QgsVectorLayerUndoPassthroughCommandAddFeatures::undo()
+{
+  if ( rollBackToSavePoint() )
+  {
+    Q_FOREACH ( const QgsFeature &f, mFeatures )
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: `Q_FOREACH` -> `for` ?

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    mFeatures = mInitialFeatures;
+  }
+}
+
+void QgsVectorLayerUndoPassthroughCommandAddFeatures::redo()
+{
+  mFeatures = mInitialFeatures;
+  if ( setSavePoint() and mBuffer->L->dataProvider()->addFeatures( mFeatures ) )
+  {
+    Q_FOREACH ( const QgsFeature &f, mFeatures )
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: `Q_FOREACH` -> `for` ?

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  : QgsVectorLayerUndoPassthroughCommand( buffer, QObject::tr( "delete features" ) )
+  , mFids( fids )
+{
+}
+
+void QgsVectorLayerUndoPassthroughCommandDeleteFeatures::undo()
+{
+  if ( rollBackToSavePoint() )
+  {
+    Q_FOREACH ( const QgsFeatureId &id, mFids )
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: `Q_FOREACH` -> `for` ?

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+      emit mBuffer->featureAdded( id );
+    }
+  }
+}
+
+void QgsVectorLayerUndoPassthroughCommandDeleteFeatures::redo()
+{
+  if ( setSavePoint() and mBuffer->L->dataProvider()->deleteFeatures( mFids ) )
+  {
+    Q_FOREACH ( const QgsFeatureId &id, mFids )
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: `Q_FOREACH` -> `for` ?

File: src/core/qgstransaction.cpp [hunk 9]
```diff
@@ -189,3 +194,54 @@ void QgsTransaction::setLayerTransactionIds( QgsTransaction *transaction )
...
+{
+  if ( !mTransactionActive )
+    return QString();
+
+  if ( !mLastSavePointIsDirty )
+    return mSavepoints.top();
+
+  const QString name( QUuid::createUuid().toString() );
+
+  if ( !executeSql( "SAVEPOINT \"" + name + "\"", error ) )
```
Comment (pblottiere) [inline src/core/qgstransaction.cpp]: `QStringLiteral().arg()` may be use here with something like `QgsExpression::quotedColumnRef` or `QgsExpression::quotedString` to manage quotes. What do you think?

Comment (pblottiere) [reply src/core/qgstransaction.cpp]: (elsewhere in the code too)

Comment (vmora) [reply src/core/qgstransaction.cpp]: Is there a good reason for that ? Because I'm not sure it improves clarity or robustness.

Comment (m-kuhn) [reply src/core/qgstransaction.cpp]: One minor advantage is less escaping for better readability. Anyway, `QStringLieteral` should be used in either case.

File: src/core/qgsvectorlayerundopassthroughcommand.h [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+#ifndef QGSVECTORLAYERUNDOPASSTHROUGHCOMMAND_H
+#define QGSVECTORLAYERUNDOPASSTHROUGHCOMMAND_H
+
+#include "qgsvectorlayerundocommand.h"
+
+#include "qgsvectorlayereditbuffer.h"
+
+/** \ingroup core
+ * \class QgsVectorLayerUndoPassthroughCommand
+ * \brief Undo command bas for vector layer in transaction group mode.
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.h]: `bas`?

File: src/core/qgsvectorlayerundopassthroughcommand.h [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+  private:
+    QString mError;
+    QString mSavePointId;
+    bool mHasError;
+    bool mRecreateSavePoint;
+
+  protected:
+
+    /** Roolback command, release savepoint or set error status
```
Comment (pblottiere) [inline src/core/qgsvectorlayerundopassthroughcommand.h]: `Rollback` instead of `Roolback`

File: src/core/qgstransaction.cpp (hunk 9 above)
Comment (vmora) [reply src/core/qgstransaction.cpp]: Ok for consistency, ok robustness, but readability:
```
QStringLiteral("SAVEPOINT %1").arg( QgsExpression::quotedString( name ) )
```

Comment (m-kuhn) [reply src/core/qgstransaction.cpp]: I just like template based solutions over concatenation ;)

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 12]
```diff
@@ -137,7 +137,7 @@ void QgsVectorLayerUndoPassthroughCommandDeleteFeatures::undo()
 {
   if ( rollBackToSavePoint() )
   {
-    Q_FOREACH ( const QgsFeatureId &id, mFids )
+    for ( const QgsFeatureId &id : mFids )
```
Comment (m-kuhn) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: I guess this should also be `qgsAsConst`?

Comment (m-kuhn) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: Forget that, all good

**Comment by vmora** [comment]
I don't understand the failure on  ProcessingQgisAlgorithmsTest (segfault), doesn't crash on my machine.

I rebased/squashed fixups to cleanup and to trigger another build.

**Comment by haubourg** [comment]
@m-kuhn Any objection for a merge?

File: src/core/qgsvectorlayerundopassthroughcommand.cpp [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+  }
+}
+
+void QgsVectorLayerUndoPassthroughCommandChangeAttribute::redo()
+{
+  QgsAttributeMap map;
+  map.insert( mField, mNewValue );
+  QgsChangedAttributesMap attribMap;
+  attribMap.insert( mFid, map );
+  if ( setSavePoint() and mBuffer->L->dataProvider()->changeAttributeValues( attribMap ) )
```
Comment (m-kuhn) [inline src/core/qgsvectorlayerundopassthroughcommand.cpp]: Can you switch `and` to`&&`?

**Comment by m-kuhn** [comment]
Would be great to quickly look through the code and find usages of operator aliases (which we nowhere else use in the codebase) and then it's ready to roll, looks like you have tested in-depth already.

Awesome stuff!

**Comment by vmora** [comment]
@m-kuhn I was not even aware of operator aliases, thanks for your vigilance, now ` git diff origin/master |grep ' and '` says I'm clean.

**Comment by vmora** [comment]
@pblottiere, @m-kuhn thanks a lot for your reviews

**Pull request merged**